### PR TITLE
New segments filters

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | n
| Related developer documentation PR URL | n
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/pull/2912
| BC breaks? | n
| Deprecations? | n

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Bunch new segment filters. Update version. More clean.
Follow up to https://github.com/mautic/mautic/pull/2912

Page filters:
- Referer
- Source
- Page Title
- Sessions  - number of visits without delay not longer like 30 minutes (same like Google Analytics visits metric)
- Visited URL Date
- Visited URL Count

Email filters
- Email sent
- Email Read Date
- Email Read Count 

Booleans filters:

URL from email
URL from landing page
Redirec URL
- check page_id, email_id, redirect_id from page_hits table 

Notifications  
- check push_id from push_ids table 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Apply and test filters

#### Steps to test this PR:
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by virlatinus** [comment]
When testing the "Email Read Count" filter and running 'console mautic:segments:update' I get an SQL error: 
'SELECT count(l.id) as lead_count, max(l.id) as max_id FROM leads l LEFT JOIN lead_lists_leads ll ON (ll.  
  leadlist_id = 14) AND (ll.lead_id = l.id) AND (ll.date_added <= '2017-02-28 13:32:18') WHERE (()) AND (ll.lead_id IS NULL)'

**Comment by kuzmany** [comment]
@virlatinus typo bug, try again pls 👍

**Review by virlatinus — APPROVED** [review]
Tested it and the PR works as described

**Comment by alanhartless** [comment]
@kuzmany we really appreciate this! We're going to push this one to 2.8 so we can test it in more depth on large data sets (think millions of page hits).

**Comment by luizeof** [comment]
There are some filter for Downloaded assets?

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 1]
```diff
@@ -906,6 +912,208 @@ public function getListFilterExpr($filters, &$parameters, QueryBuilder $q, $not
                     }
                     $groupExpr->add(sprintf('%s (%s)', $operand, $subqb->getSQL()));
                     break;
+                case 'hit_url_date':
+                case 'lead_email_read_date':
+                    $operand = ($func == 'between' || in_array($func, ['eq', 'gt', 'lt', 'gte', 'lte'])) ? 'EXISTS' : 'NOT EXISTS';
```
Comment (escopecz) [inline app/bundles/LeadBundle/Entity/LeadListRepository.php]: `($func == 'between' || in_array($func, ['eq', 'gt', 'lt', 'gte', 'lte']))`
is the same as
`(in_array($func, ['eq', 'gt', 'lt', 'gte', 'lte', 'between']))`
Isn't it?

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 2]
```diff
@@ -906,6 +912,208 @@ public function getListFilterExpr($filters, &$parameters, QueryBuilder $q, $not
...
+                    $table = 'page_hits';
+                    $column = 'date_hit';
+                   if ($details['field'] == 'lead_email_read_date') {
```
Comment (escopecz) [inline app/bundles/LeadBundle/Entity/LeadListRepository.php]: There is missing one space to align `if` with the rest of the code

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 3]
```diff
@@ -859,8 +859,15 @@ public function getListFilterExpr($filters, &$parameters, QueryBuilder $q, $not
...
                 case 'hit_url':
-                    $operand = (($func == 'eq') || ($func == 'like')) ? 'EXISTS' : 'NOT EXISTS';
-
+                case 'referer':
+                case 'source':
+                case 'url_title':
+                    $operand = in_array($func, ['eq', 'like', 'regexp' ,'notRegexp']) ? 'EXISTS' : 'NOT EXISTS';
+
+                    $column = $details['field'];
+                    if($column == 'hit_url'){
```
Comment (escopecz) [inline app/bundles/LeadBundle/Entity/LeadListRepository.php]: There are several places where the CS is not met. Could you run this PR through the pre-commit hook which will fix the CS?

**Review by escopecz — CHANGES_REQUESTED** [review]
I noticed that several small improvements could be done to the code. Please review them.

**Review by escopecz — APPROVED** [review]
I fixed the CS so this could be merged to 2.8.0. I did not find any problem during testing and the filters seems to be returning the right results 👍  Thanks!

**Pull request merged**